MNT Avoid duplicate validation of data in supervised NN

alfaro96 opened the pull request:
#### Reference Issues/PRs

https://github.com/scikit-learn/scikit-learn/pull/17847#issuecomment-654304454.

#### What does this implement/fix? Explain your changes.

This PR avoids duplicate validation of data from nearest neighbors estimators that inherits from `SupervisedIntegerMixin` and `SupervisedFloatMixin`.

#### Any other comments?

Quick review @rth for green in the CI of #17847?

alfaro96 committed:
MNT Avoid duplicate validation of data in supervised neighbors estimators

alfaro96 committed:
FIX Minor changes

rth commented on sklearn/neighbors/_base.py:
```diff
@@ -1096,15 +1094,15 @@ def fit(self, X, y):
         Parameters
         ----------
-        X : {array-like, sparse matrix, BallTree, KDTree}
+        X : {array-like, sparse matrix, BallTree, KDTree, NeighborsBase}
```
I don't understand this. When are we fitting this estimator with X=estimator?

alfaro96 replied:
This also looks a bit strange to me. I realized that we accept fitting an estimator with `X` being an estimator due to these lines:

https://github.com/scikit-learn/scikit-learn/blob/db673eb72ec2e7cbe5c30f50f97719a3bdf2f1b1/sklearn/neighbors/_base.py#L377-L382

In fact, this test:

https://github.com/scikit-learn/scikit-learn/blob/db673eb72ec2e7cbe5c30f50f97719a3bdf2f1b1/sklearn/neighbors/tests/test_neighbors.py#L98-L114

is fitting the `nbrs` estimator setting `X` to `nbrs_fid`, which is another estimator.

TomDLT replied:
This behavior is not consistent with the rest of scikit-learn's API, and it is not well documented, as briefly discussed in https://github.com/scikit-learn/scikit-learn/issues/10463#issue-288121351. I think the main goal was to avoid computing multiple times the nearest neighbors, when fitting multiple nearest-neighbors-based estimators.

Reusing precomputed nearest neighbors graphs was implemented in #10482 with an API more in line with the rest of scikit-learn. So, ideally, I would be in favor of removing the ability to use BallTree/KDTree/NeighborsBase in X, but this would need annoying deprecation cycles. We can also keep it as it is, but I would not advertise it much in the documentation.

alfaro96 replied:
> This behavior is not consistent with the rest of scikit-learn's API, and it is not well documented, as briefly discussed in [#10463 (comment)](https://github.com/scikit-learn/scikit-learn/issues/10463#issue-288121351). I think the main goal was to avoid computing multiple times the nearest neighbors, when fitting multiple nearest-neighbors-based estimators.
> 
> Reusing precomputed nearest neighbors graphs was implemented in #10482 with an API more in line with the rest of scikit-learn. So, ideally, I would be in favor of removing the ability to use BallTree/KDTree/NeighborsBase in X, but this would need annoying deprecation cycles. We can also keep it as it is, but I would not advertise it much in the documentation.

I agree with you, I would be happy to remove the ability to use `sklearn.neighbors.BallTree`, `sklearn.neighbors.KDTree` and `sklearn.neighbors.NeighborsBase` in `X` to be consistent with the API. Nevertheless, this would lead to several deprecation cycles that should be avoided.

Thus, I have removed them from the API documentation since using `sklearn.neighbors.BallTree`, `sklearn.neighbors.KDTree` and `sklearn.neighbors.NeighborsBase` in `X` is not a typical use case.

rth reviewed (commented):
Actually I find that the mixins in the NN module are not well designed. For me a mixin should add some methods, with minimal and well documented side-effects. Here that fact that we are calling `self._fit(X)` in a mixin where `_fit` is defined somewhere else, and even though the mixin does nothing with X, is confusing.  Similarly I don't think that data validation should be done in a mixin fit.

In general the fact that `KNeighborsClassifier` doesn't have a fit method but it's defined in one of 4 parent classes is not helpful IMO. Even once we found which parent defines a fit,  we  we need to find which other class defines the `_fit` method. It's very confusing.
 
As far as I can tell we are not exporting mixins to the public API so it should be fine to make this change.

Couldn't we rather add a `_validate_set_y` metod to mixin when necessary, and keep defining the fit + data validation in the main estimator?

alfaro96 commented:
> Actually I find that the mixins in the NN module are not well designed. For me a mixin should add some methods, with minimal and well documented side-effects. Here that fact that we are calling `self._fit(X)` in a mixin where `_fit` is defined somewhere else, and even though the mixin does nothing with X, is confusing. Similarly I don't think that data validation should be done in a mixin fit.
> 

> In general the fact that `KNeighborsClassifier` doesn't have a fit method but it's defined in one of 4 parent classes is not helpful IMO. Even once we found which parent defines a fit, we we need to find which other class defines the `_fit` method. It's very confusing.
> 

I agree with you. The `sklearn.neighbors.base.SupervisedIntegerMixin`, `sklearn.neighbors.base.SupervisedFloatMixin` and `sklearn.neighbors.base.UnsupervisedMixin` mixins are not well designed. They have `fit` method that is supposed to be in the `sklearn.neighbors.KNeighborsClassifier`, `sklearn.neighbors.RadiusNeighborsClassifier`, `sklearn.neighbors.KNeighborsRegressor`, `sklearn.neighbors.RadiusNeighborsRegressor` and `sklearn.neighbors.KNeighborsRegressor` estimators.

I suppose that they were intended to avoid duplicated code in the corresponding pairs of estimators. However, I think that is confusing and make a little difficult to understand the API design for the `neighbors` module.

> As far as I can tell we are not exporting mixins to the public API so it should be fine to make this change.
> 
> Couldn't we rather add a `_validate_set_y` metod to mixin when necessary, and keep defining the fit + data validation in the main estimator?

The main purpose of the `fit` method in the `sklearn.neighbors.base.SupervisedFloatMixin` and `sklearn.neighbors.base.UnsupervisedMixin` classes is to call the `_validate_data` method of the `sklearn.base.BaseEstimator`estimator class, store `y` in a private attribute (if corresponds) and then call `_fit`. Therefore, moving the `fit` + data validation in the main estimator would make these classes to disappear.

Regarding the `sklearn.neighbors.base.SupervisedIntegerMixin` class, a few more validations are applied to `y` before storing. Therefore, the `_validate_set_y` method makes sense here.

I would be happy with the `fit` + data validation in the main estimator and to have a `_validate_set_y` method to validate `y` for `sklearn.neighbors.base.SupervisedIntegerMixin`.

Another possible workaround that I have been thinking about is to have a `_validate_X_y` method in the `sklearn.neighbors.base.SupervisedFloatMixin` and `sklearn.neighbors.base.SupervisedIntegerMixin` classes and `_validate_X` in `sklearn.neighbors.base.UnsupervisedMixin`.

WDYT @rth?

rth commented:
> I would be happy with the fit + data validation in the main estimator and to have a _validate_set_y method to validate y for sklearn.neighbors.base.SupervisedIntegerMixin.

+1 to try that.

> Another possible workaround that I have been thinking about is to have a _validate_X_y method in the sklearn.neighbors.base.SupervisedFloatMixin and sklearn.neighbors.base.SupervisedIntegerMixin classes and _validate_X in sklearn.neighbors.base.UnsupervisedMixin. By this way, we will not have to duplicate code in sklearn.neighbors.KNeighbors* and sklearn.neighbors.RadiusNeighbors*.

Just in terms of method name I find `_validate_X_y` a bit confusing with `BaseEstimator._validate_data(X, y)` that we are supposed to use everywhere. Maybe let's start with the version that includes those small code duplication, and then see if they need to be factorized further?  Or up to you, I haven't studied this module code in detail, so you probably have a better vision of what can be done there.

alfaro96 commented:
> > I would be happy with the fit + data validation in the main estimator and to have a _validate_set_y method to validate y for sklearn.neighbors.base.SupervisedIntegerMixin.
> 
> +1 to try that.
> 
> > Another possible workaround that I have been thinking about is to have a _validate_X_y method in the sklearn.neighbors.base.SupervisedFloatMixin and sklearn.neighbors.base.SupervisedIntegerMixin classes and _validate_X in sklearn.neighbors.base.UnsupervisedMixin. By this way, we will not have to duplicate code in sklearn.neighbors.KNeighbors* and sklearn.neighbors.RadiusNeighbors*.
> 
> Just in terms of method name I find `_validate_X_y` a bit confusing with `BaseEstimator._validate_data(X, y)` that we are supposed to use everywhere. Maybe let's start with the version that includes those small code duplication, and then see if they need to be factorized further? Or up to you, I haven't studied this module code in detail, so you probably have a better vision of what can be done there.

I will focus on the `_validate_set_y` version and then let us see if we can refactor some code.

alfaro96 commented:
I think that I should apply these changes in this PR to avoid a long dependence chain among PRs.

WDYT @rth?

rth commented:
Yes, sounds good.

alfaro96 committed:
MNT Remove fit method from neighbors mixins

alfaro96 commented on sklearn/neighbors/_base.py:
```diff
@@ -1155,20 +1120,9 @@ def fit(self, X, y):
-            Training data. If array or matrix, shape [n_samples, n_features],
-            or [n_samples, n_samples] if metric='precomputed'.
-        """
-        return self._fit(X)
+    pass
```
Note that I have kept this class empty to avoid merge conflicts in #17847.

alfaro96 committed:
MNT Minor changes

rth reviewed (approved):
I find this much easier to understand and it LGTM. Repetition of `fit` is not really an issue IMO.

Do we need to keep `UnsupervisedMixin` if it does nothing?

WDYT (about this change in general) @TomDLT ?

alfaro96 commented:
> I find this much easier to understand and it LGTM. Repetition of `fit` is not really an issue IMO.
> 

I agree with you, the `neighbors` module is much easier to understand right now 😄.

> Do we need to keep `UnsupervisedMixin` if it does nothing?
> 

I do not think that we need to keep `sklearn.neighbors.base.UnsupervisedMixin`, but I have kept to avoid merge conflicts in #17847 (for the MRO).

> WDYT (about this change in general) @TomDLT ?

Thanks @rth for the quick review!

rth commented:
> Nevertheless, do you think that we should remove in this PR and solve the merge conflicts 

Yeah, if it's not useful, let's do that. We could always solve conflicts once this PR is merged.

alfaro96 committed:
MNT Remove UnsupervisedMixin from the private API

TomDLT reviewed (approved):
I like this change.

Are we ok removing mixins like that ? Are they part of the public API in some way ?

rth commented:
> Are we ok removing mixins like that ? Are they part of the public API in some way ?

They are only used in private modules as parents of public estimators. Technically it is a private mixin, so it's probably fine.

glemaitre commented on sklearn/neighbors/_classification.py:
```diff
@@ -159,6 +161,32 @@ def __init__(self, n_neighbors=5, *,
+        self : KNeighborsClassifier
+            The fitted k-nearest neighbors classifier.
+        """
+        if not isinstance(X, (KDTree, BallTree, KNeighborsClassifier)):
```
This would need a test then.

alfaro96 replied:
I have applied the suggested changes and included the corresponding tests.

Thanks @TomDLT and @glemaitre for the suggestions!

glemaitre reviewed (commented):
I should admit that I had to read the code to understand why we had these mixins. What confused me was the naming.

```python
class _SupervisedMixin:
    def _more_tags(self):
        return {requires_y': True}

class _SupervisedRegressorMixin(_SupervisedMixin):
    """For completness"""

class _SupervisedClassifierMixin(_SupervisedMixin):
    def _validate_classification_targets(self, y):
        ...
```

glemaitre commented:
Actually my proposal is bad because I make the Mixin inherit from something.

One thing that I find weird is that `ClassifierMixin` and `RegressorMixin` already define the `"require_y": True` so we should not need the `_Supervised*` base class. Then, the check on the classification target, I would expect it to be implemented in the `ClassifierMixin` directly to be shared across classifier, isn't it?

alfaro96 committed:
CLN Apply suggested changes

alfaro96 commented:
> Actually my proposal is bad because I make the Mixin inherit from something.
> 

I agree with that, I do not really like to have `_SupervisedRegressorMixin` and `_SupervisedClassifierMixin` inherit from `_SupervisedMixin`.

> One thing that I find weird is that `ClassifierMixin` and `RegressorMixin` already define the `"require_y": True` so we should not need the `_Supervised*` base class. Then, the check on the classification target, I would expect it to be implemented in the `ClassifierMixin` directly to be shared across classifier, isn't it?
> 

I am +1 about having `check_classification_targets` shared across all classifiers, I pretty like this solution. Maybe we can include this check within `_validate_data`?

> And what makes the specificity of the Neighbors algorithm is to store the `y` in `self._y` but it could be done in `_fit` and not in the validation of the `y`

I am wondering that we could change the signature of `_fit` from `_fit(self, X)` to `_fit(self, X, y=None)` and move these lines:

```python
if not isinstance(X, (KDTree, BallTree, NeighborsBase)):
    X, y = self._validate_data(X, y, accept_sparse="csr",
                               multi_output=True)
```

```python
if not isinstance(X, (KDTree, BallTree, NeighborsBase)):
    X = self._validate_data(X, accept_sparse='csr')
```

inside the `_fit` method, using the `require_y` tag to know whether to validate `y`.

Moreover, the `_validate_set_y` method can be moved to `NeighborsBase` class and, using the `is_classifier` function, we should know whether to call `_validate_set_y`.

WDYT?

TomDLT commented:
I agree these mixins should not exist in the first place, the validation of `y` can be shared with all classifiers (it is probably already implemented somewhere else), and the call to `_validate_data` can be moved to `_fit`.

alfaro96 committed:
MNT More _validate_data to _fit

alfaro96 commented:
I think that I have already addressed all the points to tackle. I am happy to see a clearer private API in the `sklearn.neighbors` module 😄

**Note: I have also modified the MRO for the nearest neighbors estimators to remove all unnecessary mixins. I would be happy to close #17847 without merging if you all agree.**

rth commented on sklearn/neighbors/_classification.py:
```diff
@@ -401,22 +422,25 @@ def __init__(self, radius=1.0, *, weights='uniform',
 
     def fit(self, X, y):
-        """Fit the model using X as training data and y as target values
+        """Fit the radius neighbors classifier using `X` as training data and
+        `y` as target values.
```
This should be a 1 line summary: https://numpydoc.readthedocs.io/en/latest/format.html#sections
Also apparently that's does not use variable names, but I think the previous version was fine.

I think the previous version was fine.

alfaro96 replied:
Thanks @rth for the suggestion!

I have kept the 1-line summary and remove the variable names.

rth commented on sklearn/neighbors/_regression.py:
```diff
@@ -160,6 +160,27 @@ def _pairwise(self):
 
+    def fit(self, X, y):
+        """Fit the k-nearest neighbors regressor using `X` as training data
+        and `y` as target values.
```
Also should be a 1 line summary.

rth commented on sklearn/neighbors/_regression.py:
```diff
@@ -328,6 +349,27 @@ def __init__(self, radius=1.0, *, weights='uniform',
 
+    def fit(self, X, y):
+        """Fit the radius neighbors regressor using `X` as training data and
+        `y` as target values.
```
Same.

rth reviewed (commented):
Thanks @alfaro96 ! A few minor comments otherwise LGTM.

rth commented on sklearn/neighbors/_classification.py:
```diff
@@ -159,6 +158,27 @@ def __init__(self, n_neighbors=5, *,
 
+    def fit(self, X, y):
+        """Fit the k-nearest neighbors classifier using `X` as training data
+        and `y` as target values.
```
Same.

alfaro96 committed:
CLN Apply suggested changes

alfaro96 commented:
I have applied the suggested changes. Let us wait for the green in the CI 😄

rth commented:
Thanks @alfaro96 !

glemaitre commented:
Thanks @alfaro96 It looks better I think :)
